File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/Quit_Notification.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import com.google.common.base.Strings;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;

@AutoValue
public abstract class Quit_Notification extends Quit {
  private static final String DESC = "User Quit Notification";

  public static AutoValue_Quit_Notification create(
      int messageNumber, String username, int userId, String message)
      throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (userId < 0 || userId > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userID out of acceptable range: " + userId);
    }

    if (message == null)
      throw new MessageFormatException("Invalid " + DESC + " format: message == null!");

    if (Strings.isNullOrEmpty(username)) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userName.length == 0, (userID = " + userId + ")");
    }
    return new AutoValue_Quit_Notification(messageNumber, ID, DESC, username, userId, message);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/Quit_Notification.kt
package org.emulinker.kaillera.controller.v086.protocol

import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber

data class Quit_Notification
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int,
        override val username: String,
        override val userId: Int,
        override val message: String
    ) : Quit() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
    require(userId in 0..0xFFFF) { "UserID out of acceptable range: $userId" }
    require(username.isNotBlank()) { "Username cannot be empty" }
  }
}
